diskquota_utility.c

```diff
@@ -1190,3 +1192,80 @@ get_rel_oid_list(void)
 	}
 	return oidlist;
 }
+
+/*
+ * calculate size of (all forks of) a relation in transaction
+ * This function is following calculate_relation_size()
+ */
+static Size
+calculate_relation_size_all_forks(RelFileNodeBackend *rnode)
+{
+    int64       totalsize = 0;
+    ForkNumber  forkNum;
+    int64       size = 0;
+    char       *relationpath;
+    char        pathname[MAXPGPATH];
+    unsigned int segcount = 0;
+    PG_TRY();
+	{
+        for (forkNum = 0; forkNum <= MAX_FORKNUM; forkNum++)
+        {
+            relationpath = relpathbackend(rnode->node, rnode->backend, forkNum);
+            size = 0;
+
+            for (segcount = 0;; segcount++)
+            {
+                struct stat fst;
+
+                CHECK_FOR_INTERRUPTS();
+
+                if (segcount == 0)
+                    snprintf(pathname, MAXPGPATH, "%s",
+                            relationpath);
+                else
+                    snprintf(pathname, MAXPGPATH, "%s.%u",
+                            relationpath, segcount);
+
+                if (stat(pathname, &fst) < 0)
+                {
+                    if (errno == ENOENT)
+                        break;
+                }
+                size += fst.st_size;
+            }
+
+            totalsize += size;
+        }
+    }
+	PG_CATCH();
+	{
+		// TODO: Record the error message to pg_log
+		HOLD_INTERRUPTS();
+		FlushErrorState();
+		RESUME_INTERRUPTS();
+	}
+	PG_END_TRY();
+
+    return totalsize;
+}
+
+PG_FUNCTION_INFO_V1(relation_size_local);
+
+Datum
+relation_size_local(PG_FUNCTION_ARGS)
+{
+    Oid reltablespace = PG_GETARG_OID(0);
+	Oid relfilenode = PG_GETARG_OID(1);
+    int backend = PG_GETARG_BOOL(2) ? -2 : -1;
+    RelFileNodeBackend rnode = {0};
+    Size size = 0;
+
+    rnode.node.dbNode = MyDatabaseId;
+    rnode.node.relNode = relfilenode;
+    rnode.node.spcNode = reltablespace ? reltablespace : DEFAULTTABLESPACE_OID;
```

Comment:
```suggestion
    rnode.node.spcNode = OidIsValid(reltablespace) ? reltablespace : DEFAULTTABLESPACE_OID;
```

Though C compilers can do implicit type conversion for us, I still recommend using `OidIsValid()`.